pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -159,4 +181,271 @@ def is_root_lvl(patch_level):
+               "flux_z": "Vz",
+               "bulkVelocity_x": "Vx",
+               "bulkVelocity_y": "Vy",
+               "bulkVelocity_z": "Vz",
+               "density": "rho"}
+
+
+particle_files_patterns = ("domain", "patchGhost", "levelGhost")
+
+
+def is_particle_file(filename):
+    return any([pattern in filename for pattern in particle_files_patterns])
+
+
+
+def particle_dataset_name(basename):
+    """
+    return "alpha_domain" from ions_pop_ions_alpha_domain.h5
+    """
+    popname = basename.strip(".h5").split("_")[-2]
+    part_type = basename.strip(".h5").split("_")[-1]
+    dataset_name = popname + "_" + part_type
+
+    return dataset_name
+
+
+def is_pop_fluid_file(basename):
+    return (is_particle_file(basename) is False) and "pop" in basename
+
+
+def pop_name(basename):
+    return basename.strip(".h5").split("_")[-2]
+
+
+def make_layout(h5_patch_grp, cell_width):
+    nbrCells = h5_patch_grp.attrs['nbrCells']
+    origin = float(h5_patch_grp.attrs['origin'])
+    upper = int(h5_patch_grp.attrs['upper'])
+    lower = int(h5_patch_grp.attrs['lower'])
+
+    return GridLayout(Box(lower, upper), origin, cell_width)
+
+
+
+def create_from_all_times(time, hier):
+    return time is None and hier is None
+
+
+def create_from_one_time(time, hier):
+    return time is not None and hier is None
+
+
+def load_all_times(time, hier):
+    return time is None and hier is not None
+
+
+def load_one_time(time, hier):
+    return time is not None and hier is not None
+
+
+
+
+def add_to_patchdata(patch_datas, h5_patch_grp, basename, layout):
+    """
+    adds data in the h5_patch_grp in the given PatchData dict
+    """
+
+    if is_particle_file(basename):
+        particles = Particles(icells=h5_patch_grp["iCell"],
+                              deltas=h5_patch_grp["delta"],
+                              v=h5_patch_grp["v"],
+                              weights=h5_patch_grp["weight"],
+                              charges=h5_patch_grp["charge"])
+
+        pdname = particle_dataset_name(basename)
+        if pdname in patch_datas:
+            raise ValueError("error - {} already in patchdata".format(pdname))
+
+        patch_datas[pdname] = ParticleData(layout, particles)
+
+    else:
+        for dataset_name in h5_patch_grp.keys():
+
+            dataset = h5_patch_grp[dataset_name]
+
+            if dataset_name not in field_qties:
+                raise RuntimeError(
+                    "invalid dataset name : {} is not in {}".format(dataset_name, field_qties))
+
+            pdata = FieldData(layout, field_qties[dataset_name], dataset)
+
+            if is_pop_fluid_file(basename):
+                dataset_name = pop_name(basename) + "_" + dataset_name
+
+
+            if dataset_name in patch_datas:
+                raise ValueError("error - {} already in patchdata".format(dataset_name))
+
+            patch_datas[dataset_name] = pdata
+
+
+
+
+
+
+def hierarchy_from(h5_filename, time=None, hier=None):
```

Comment:
Why this format vs
```
hier.add(file, time) 
```

?

Reply:
mainly, hier.add() means `add` is a method of PatchHierarchy, and I think it's not the role of the hierarchy to know about h5 files.

Reply:
some type asserts would make me happier given "hierarchy_from" can accept anything as parameters

Reply:
> and I think it's not the role of the hierarchy to know about h5 files.

I don't buy this.
there's an implicit relationship regardless.
the PH doesn't have to handle any input directly, but it has no purpose other than to accept information from somewhere

Reply:
yeah the dependency of the PH on h5 is implicit at at runtime, typically when datasets get assigned in patchDatas they are assigned to h5 datasets, and when [:] is called, h5 code is called. But there is no need of explicit "h5" code related to how the h5 file is structured, in the PH , that's what I meant. This way, upon changes in the file format, the PH code stays as is.